File: StrandedDeepAllInOneTool/StrandedDeepMapper/ZonePositionGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace StrandedDeepMapper
{
    public static class ZonePositionGenerator
    {
        public static Vector2[] GeneratePoints(int WORLD_SEED, double radius, Vector2 sampleRegionSize, int numSamplesBeforeRejection = 30)
        {
            //string path = @"FastRandomLog.txt";

            double num = radius / Math.Sqrt(2f);
            //int[,] array = new int[Math.CeilToInt(sampleRegionSize.x / num), Math.CeilToInt(sampleRegionSize.y / num)];
            int[,] array = new int[(int)Math.Ceiling(sampleRegionSize.x / num), (int)Math.Ceiling(sampleRegionSize.y / num)];
            List<Vector2> list = new List<Vector2>();
            List<Vector2> list2 = new List<Vector2>();
            list2.Add(sampleRegionSize / 2f);
            //FastRandom fastRandom = new FastRandom(WORLD_SEED);
            FastRandomNew fastRandom = new FastRandomNew(WORLD_SEED);
            while (list2.Count > 0)
            {
                int index = fastRandom.Next(0, list2.Count);
                Vector2 vector = list2[index];
                bool flag = false;
                for (int i = 0; i < numSamplesBeforeRejection; i++)
                {
                    double randomValue = fastRandom.NextDouble();

                    float test = (float)randomValue;
                    float test2 = (float)Math.PI;
                    float test3 = test * test2 * 2f;

                    float f = (float)randomValue * (float)Math.PI * 2f;
                    Vector2 a = new Vector2(Math.Sin(f), Math.Cos(f));
                    int nextRandom = fastRandom.Next((int)(radius * 100f), (int)(1.25f * radius * 100f));
                    Vector2 vectortemp = a * (nextRandom / 100);
                    Vector2 vector2 = vector + vectortemp;

                    //using (StreamWriter sw = File.AppendText(path))
                    //{
                    //    sw.WriteLine("sin/cos$" + Math.Sin(f) + "$" + Math.Cos(f));
                    //    sw.WriteLine(randomValue + "$" + f + "$" + a.x + "$" + a.y + "$" + nextRandom + "$" + vector2.x + "$" + vector2.y);
                    //}

                    if (i == 0)
                    {
                        vector2 = vector;
                    }
                    bool isValid = IsValid(vector2, sampleRegionSize, num, radius, list, array);
                    //using (StreamWriter sw = File.AppendText(path))
                    //{
                    //    sw.WriteLine("IsValid " + vector2.x + "/" + vector2.y + " /" + isValid);
                    //}
                    if (isValid)
                    {
                        list.Add(vector2);
                        list2.Add(vector2);
                        array[(int)(vector2.x / num), (int)(vector2.y / num)] = list.Count;
                        flag = true;
                        break;
                    }
                }
                if (!flag)
                {
                    list2.RemoveAt(index);
                }

                //using (StreamWriter sw = File.AppendText(path))
                //{
                //    sw.WriteLine("list2 length = " + list2.Count);
                //}
            }
            Vector2[] array2 = new Vector2[list.Count];
            for (int j = 0; j < list.Count; j++)
            {
                double num2 = sampleRegionSize.x * 0.5f;
                ref Vector2 reference = ref array2[j];
                Vector2 vector3 = list[j];
                double x = vector3.x - num2;
                Vector2 vector4 = list[j];
                reference = new Vector2(x, vector4.y - num2);
            }
            return array2;
        }

        private static bool IsValid(Vector2 candidate, Vector2 sampleRegionSize, double cellSize, double radius, List<Vector2> points, int[,] grid)
        {
            //string path = @"FastRandomLog.txt";
            //using (StreamWriter sw = File.AppendText(path))
            //{
                if (candidate.x >= 0f && candidate.x < sampleRegionSize.x && candidate.y >= 0f && candidate.y < sampleRegionSize.y)
                {
                    int num = (int)(candidate.x / cellSize);
                    int num2 = (int)(candidate.y / cellSize);
                    int num3 = Math.Max(0, num - 2);
                    int num4 = Math.Min(num + 2, grid.GetLength(0) - 1);
                    int num5 = Math.Max(0, num2 - 2);
                    int num6 = Math.Min(num2 + 2, grid.GetLength(1) - 1);
                    //sw.WriteLine("IsValid : num " + num + " / num2 " + num2 + " / num3 " + num3 + " / num4 " + num4 + " / num5 " + num5 + " / num6 " + num6);
                    for (int i = num3; i <= num4; i++)
                    {
                        for (int j = num5; j <= num6; j++)
                        {
                            int num7 = grid[i, j] - 1;
                            if (num7 != -1)
                            {
                                double sqrMagnitude = (candidate - points[num7]).sqrMagnitude;
                                //sw.WriteLine("sqmagnitude = " + sqrMagnitude.ToString("G17") + " < radius2 = " + (radius * radius) + "$" + points[num7].x + "$" + points[num7].y + "$" + candidate.x + "$" + candidate.y);
                                if (sqrMagnitude < radius * radius)
                                {
                                    //sw.WriteLine("IsValid : " + candidate.x + " / " + candidate.y + " / false (sqmagnitude=" + sqrMagnitude.ToString("G17") + " < radius2=" + (radius * radius) + ")");
                                    return false;
                                }
                            }
                        }
                    }
                    //sw.WriteLine("IsValid : " + candidate.x + " / " + candidate.y + " true ");
                    return true;
                }
                //sw.WriteLine("IsValid : " + candidate.x + " / " + candidate.y + " / false (sampleRegionSize.x=" + sampleRegionSize.x + " / sampleRegionSize.y=" + sampleRegionSize.y + ")");
                return false;
            }
        //}
    }
}

File: Stranded Deep JS Mapper - obsolete/SDmapperJS - 0.17/zonepositiongenerator.js
// right now i'm very close to giving up
class ZonePositionGenerator {

	GeneratePoints(WORLD_SEED, radius, sampleRegionSize, numSamplesBeforeRejection = 30) {
		let num = radius / Math.sqrt(2);
		let myarray = new Array();//new int[Math.ceil(sampleRegionSize.x / num), Math.ceil(sampleRegionSize.y / num)];
		// JS specific init loop
		var x, y;
		for(x = 0; x < Math.ceil(sampleRegionSize.x / num); x++)
		{
			myarray.push(new Array());
			for(y = 0; y < Math.ceil(sampleRegionSize.y / num); y++)
			{
				myarray[x].push(0);
			}
		}
		
		let mylist = new Array();//new List<Vector2>();
        let mylist2 = new Array();//new List<Vector2>();
		mylist2.push(sampleRegionSize.divide(2));
		let fastRandom = new FastRandom(WORLD_SEED);
		while (mylist2.length > 0)
        {
			let index = fastRandom.Next(0, mylist2.length);
			let vector = mylist2[index];
			let flag = false;
			var i;
			for (i = 0; i < numSamplesBeforeRejection; i++)
            {
				let randomvalue = fastRandom.NextDouble();
				let f = (randomvalue.toFixed(8) * Math.PI.toFixed(8) * 2).toFixed(8);
				let a = new Vector2(Math.sin(f), Math.cos(f));
				let nextRandom = fastRandom.Next(Math.trunc(radius * 100), Math.trunc(1.25 * radius * 100));
				let vectortemp = a.multiply(nextRandom / 100);
                let vector2 = vector.addXY(vectortemp.x, vectortemp.y);
				if (i == 0)
                {
					vector2 = vector;
                }
				if (this.IsValid(vector2, sampleRegionSize, num, radius, mylist, myarray))
                {
                    mylist.push(vector2);
                    mylist2.push(vector2);
                    myarray[Math.trunc(vector2.x / num)][Math.trunc(vector2.y / num)] = mylist.length;
                    flag = true;
                    break;
                }
			}
			if (!flag)
            {
				mylist2.splice(index, 1);
            }
			//console.log("list2 length = " + mylist2.length);
		}
		let array2 = Array();//new Vector2[list.Count];
		// JS specific init loop
		var i;
		for (i = 0; i < mylist.length; i++)
		{
			array2.push(new Vector2(0, 0));
		}
		
		var j;
        for (j = 0; j < mylist.length; j++)
        {
            let num2 = sampleRegionSize.x * 0.5;
            // ref Vector2 reference = ref array2[j];
            let vector3 = mylist[j];
            let x = vector3.x - num2;
            let vector4 = mylist[j];
            // reference = new Vector2(x, vector4.y - num2);
			array2[j] = new Vector2(x, vector4.y - num2);
        }
        return array2;
	}
	

	IsValid(candidate, sampleRegionSize, cellSize, radius, points, mygrid)
	{
		if (candidate.x >= 0 && candidate.x < (sampleRegionSize.x + 0.999999) && candidate.y >= 0 && candidate.y < (sampleRegionSize.y + 0.999999)) // error margin due to JS + 0.999999
		{
			let num = Math.trunc(candidate.x / cellSize);
			let num2 = Math.trunc(candidate.y / cellSize);
			let num3 = Math.max(0, num - 2);
			let num4 = Math.min(num + 2, mygrid.length - 1);
			let num5 = Math.max(0, num2 - 2);
			let num6 = Math.min(num2 + 2, mygrid[0].length - 1);
			var i, j;
			for (i = num3; i <= num4; i++)
			{
				for (j = num5; j <= num6; j++)
				{
					let num7 = mygrid[i][j] - 1;
					if (num7 != -1)
					{
						let sqrMagnitude = candidate.substractXY(points[num7].x, points[num7].y).sqrMagnitude();
						if (sqrMagnitude < (radius * radius)) // error margin due to JS  * 0.9995
						{
							//console.log("IsValid : x=" + candidate.x + " / y=" + candidate.y + " / false (sqmagnitude="+sqrMagnitude+" / radius2="+radius * radius+")");
							return false;
						}
					}
				}
			}
			//console.log("IsValid : x=" + candidate.x + " / y=" + candidate.y + " / true");
			return true;
		}
		//console.log("IsValid : x=" + candidate.x + " / y=" + candidate.y + " / false (sampleRegionSize.x="+sampleRegionSize.x+" / sampleRegionSize.y="+sampleRegionSize.y+")");
		return false;
	}
}